core/message-tags-3.3.md

```diff
@@ -111,6 +117,22 @@ See [`PRIVMSG` in RFC2812](https://tools.ietf.org/html/rfc2812#section-3.3.1) fo
 
 Clients that receive a `TAGMSG` command MUST NOT display them in the message history by default. Display guidelines are defined in the specifications of tags attached to the message.
 
+### JSON encoding format
+
+To send tags as JSON, clients and servers MUST encode tag data as an *object* whos top-level keys are the tag keys, encoded as *strings*
+
+Tag keys without a value MUST be represented in JSON as keys with a `null` value.
+
+Tag values MUST be encoded as JSON *string* types, except where defined in specific tag specifications.
+
+All extraneous whitespace MUST be removed from encoded JSON, and the entire encoded JSON data MUST be escaped according to existing tag value escaping rules.
+
+Fully encoded and escaped JSON tag data is transmitted instead of `<tags>` in the originally defined pseudo-BNF.
+
+To detect JSON-encoded tag data, clients and servers MUST check for `'{'` (0x7B) as the first character of tag data, after the leading `'@'` (0x40) character. The JSON data ends on the first space `' '` (0x20) character.
+
+If JSON tag data fails to decode correctly, clients MUST abort message parsing and discard the entire message, as this may be a symptom of incorrectly applied escaping rules, making the remainder of the message vulnerable to command injection. Servers MUST ignore incorrectly encoded JSON data entirely and MAY terminate the client connection as a result.
```

Comment:
Anyone have any comments on this? Technically, the injection risk here would also be an issue for incorrectly implemented standard tag encoding, but there isn't really a way to detect it in the same way as with a json parse error.

Also, should the server behaviour be defined more strictly here? Should there be an error numeric for invalid tag data? Should the connection terminate option be a MUST? Should we allow servers to accept the message but just ignore the tag part?

Reply:
Making the termination "MUST" doesn't make much sense to me. IRC discards invalid messages right now as well, so why should we make it terminating the connection?

Ignoring parts of a message doesn't make much sense to me as well, because what if the tag is necessary? If we can't get the entire context of something, we shouldn't do anything with it. Otherwise, we may run into unneeded security issues.

Reply:
So do you think the server should silently ignore the whole message or respond with an error numeric?

Reply:
I would go for a numeric. When we think about regular IRC messages right now, you would get `421 ERR_UNKNOWNCOMMAND` or similar. So we should maybe provide one for "invalid message format"?

Reply:
I don't think an existing numeric exists that could be used for this. Should we define a text error response as have started to be proposed like `ERR` and `WARN`in e.g. chathistory #292? The difference being the chathistory error responses are scoped by a namespace.

These unresolved questions are partly why I added the "may terminate" option. Also, if a client is sending clearly badly encoded junk at this level of the protocol, it's probably not recoverable, so an error response is of questionable use.

That said, a simple solution could just be to reuse `412 ERR_UNKNOWNCOMMAND` although it's not necessarily the most descriptive or accurate. Possibly with a more descriptive message parameter than `Unknown command`?

Reply:
Oh, the 412 was only an example how a similar event is handled by RFC1459.

It wasn't intended to use it for this event. I would go for an own numeric here. If it should be namespaced or not, I'm not sure.

The "MAY terminate" is fine. But I wouldn't force it. This would possibly lead to DOS scenarios where people send a good request to another client, that replies with broken client tags and get disconnected.

Reply:
Yes I understood your comment, but we don't really have a good solution for numeric exhaustion right now so there needs to be a **really** strong reason to define new one. That's why I think it would be the easiest to just reuse `421` here, it's what servers already do when you try to feed them invalid stuff.

I'm not particularly concerned about DOS scenarios when this would be a symptom of a fatally broken client.

Reply:
All of that said, going back to your previous statement against just ignoring the tag part:

>Ignoring parts of a message doesn't make much sense to me as well, because what if the tag is necessary? If we can't get the entire context of something, we shouldn't do anything with it. Otherwise, we may run into unneeded security issues.

Isn't that already a problem for the standard tag encoding? If you make an error encoding tags, current server implementations will just ignore the invalid tag data, even if it's "necessary", so do we really need to be more strict for json tags?

Reply:
Decided to just change this to let invalid json be ignored entirely. Significantly simplifies implementations and avoids a few of the issues mentioned here.